Add Particle Cloud integration

Schoonology opened the pull request:
## Description:

Adds a new integration: Particle Cloud

This follows the implementation guideline to use an external library to communicate with the external service, and documents a new service, `particle.call`, used to call functions on the Particle devices. Additionally all variables published from associated devices are made available as device attributes.

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#11085

## Example entry for `configuration.yaml` (if applicable):
```yaml
particle:
  access_token: !secret particle_cloud_token
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

Schoonology committed:
Add integration with the Particle Cloud.

Schoonology committed:
Add Particle integration library.

Schoonology committed:
Exclude Particle implementation from code coverage.

Schoonology commented:
It looks like there are still items to work on here (my local dev environment stopped noticing where `tox` is located...), but I'd appreciate any and all initial feedback.

I had posted for a bit of feedback in Discord, but didn't hear anything yet, so decided to push what I have anyway. 😄

fabaff commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,100 @@
+_LOGGER = logging.getLogger(DOMAIN)
+CONFIG_SCHEMA = vol.Schema({
+  DOMAIN: vol.Schema({
+    vol.Required('access_token'): cv.string,
```
We have `CONF_ACCESS_TOKEN` in `const.py`. Please import that.

fabaff commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,100 @@
+    function = service.data.get('function')
+    args = service.data.get('args', [])
+
+    source = 'self._info.' + function + '(*args)'
```
Also, string formatting.

fabaff commented on homeassistant/components/particle/manifest.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "domain": "particle",
+  "name": "Particle Cloud",
+  "documentation": "",
```
```suggestion
  "documentation": "https://www.home-assistant.io/integrations/particle/",
```

fabaff commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,100 @@
+}, extra=vol.ALLOW_EXTRA)
+
+FUNCTION_CALL_SCHEMA = ENTITY_SERVICE_SCHEMA.extend({
+  vol.Required('function'): cv.string,
```
Use a constant here.

```suggestion
  vol.Optional(ATTR_FUNCTION): vol.All(
```

along with `ATTR_FUNCTION = 'function'`

fabaff commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,100 @@
+  devices = {}
+  device_list = []
+
+  for name, info in particle.devices.items():
```
I assume that if the access token is not valid that there will be a traceback. You should check, if possible, that the integration works during the setup process and make it fails if not. There is no way to recover if the credentials are wrong.

fabaff commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,100 @@
+  ),
+})
+
+def setup(hass, config):
```
Should be async.

fabaff reviewed (commented):
Some comments to get the review started...

Schoonology committed:
Update the Particle component based on feedback.

- Added docstrings.
- Ran `isort` on dependencies.
- Updated formatting, particularly of strings.
- Setup is now async.
- Errors are now caught and logged.
- Linked to documentation.

Schoonology commented:
Pushed a pull request to update the documentation: https://github.com/home-assistant/home-assistant.io/pull/11085

Schoonology commented:
Thank you, @fabaff, for your very generous review. I updated the integration accordingly.

fabaff commented:
There is a merge conflict.

Schoonology committed:
Merge branch 'dev' into component-particle

Trekky12 commented:
Hey @Schoonology, 

apparently there are still errors when running the CI checks. 
Could you run `python3 -m script.hassfest` to update the CODEOWNERS?

Thank you very much! I'am already excited to use this integration 👍

Schoonology committed:
Update CODEOWNERS for Particle integration.

Schoonology commented:
@Trekky12 I don't see the bandit tool—which errored in CI—mentioned in the docs anywhere, but I might be missing something. Do I need to rebase or otherwise update this PR?

Trekky12 commented:
@Schoonology You can see here in the PR the failed checks. You already successfully fixed the CODEOWNERS file but unfortunately the CI is still showing an error:

As you can see in the [details](https://dev.azure.com/home-assistant/Core/_build/results?buildId=25861&view=logs&j=4e214fed-f310-53fa-0018-b38e7beb405b&t=b826438c-e413-57c7-174a-d3bdede8f334&l=29), [bandit](https://github.com/PyCQA/bandit) is exiting with an error since you are using the `eval` function. Could you change this accordingly and use `ast.literal_eval` like recommended?

Schoonology committed:
Particle: Remove unnecessary eval.

Schoonology commented:
Ah. Missed that looking through the details initially. Looks like bandit's output gets garbled a little bit in CI.

Either way, while I previously thought this code _required_ the use of `eval` over `ast.literal_eval` (it makes function calls), I think it's just been too long since I wrote Python on a daily basis: I'm calling `__getattr__` directly, which should obviate the need for `eval` in the first place.

I did not design the API for `pyparticleio`, which works great for scripting, but less so for proxying another, more flexible API on top. 🤷‍♂

Schoonology commented:
To that end, one of the current lint failures is directly tied to the design of the upstream integration: if there is an error while logging in, `pyparticleio` raises a broad `Exception`. The broad `except` statement is caught as lint; what's the HA-blessed route to proceed?

Schoonology committed:
Migrate to make_entity_service_schema.

Side effect: Fix lint error.

Schoonology committed:
Particle: Ignore `broad-except` where required.

Schoonology commented:
Looks like there's some precedent for using `# pylint: disable=broad-except`, so I added that for now.

Schoonology commented:
@Trekky12 Is it common for folks to merge their own PRs? I'm used to projects where existing contributors merge, to have more fine-grained control over when patches land.

Trekky12 commented:
I'm not that familiar with the process of merging the commits, maybe @cgarwood can have a look?

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    if DOMAIN not in config:
+        return True
+
+    from pyparticleio.ParticleCloud import ParticleCloud
```
Please move all imports to the top.

Schoonology replied:
So, let's talk about this. As far as earlier reviews were concerned, this PR was ready to go. Is there a written-down process for getting these merged?

The style guide explicitly requests modules only be loaded when required, which is why the `import` statement is here. What about conflicts between these additional comments, and the style?

I'd like clarity on where we go from here before I put more time into this code.

Thanks.

frenck replied:
That rule has been lifted quite some time ago, the style guideline requiring the import when it was used, has been removed at that time as well.

Please move the imports to the top (as we did quite a big effort to move everything to the top for the hundreds of integrations we have 😉)

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+
+    try:
+        particle = hass.data[DOMAIN] = ParticleCloud(access_token)
+    except Exception as ex:  # pylint: disable=broad-except
```
Only catch the specifc exceptions that could occur and remove `# pylint: disable=broad-except`.

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    try:
+        particle = hass.data[DOMAIN] = ParticleCloud(access_token)
+    except Exception as ex:  # pylint: disable=broad-except
+        _LOGGER.error("Unable to connect to Particle Cloud: %s", str(ex))
```
```suggestion
        _LOGGER.error("Unable to connect to Particle Cloud: %s", ex)
```

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+                try:
+                    self._attributes[name] = self._info.__getattr__(name)
+                except Exception as ex:  # pylint: disable=broad-except
+                    _LOGGER.error("Unable to update from Particle Cloud: %s", str(ex))
```
```suggestion
                    _LOGGER.error("Unable to update from Particle Cloud: %s", ex)
```

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+            for name, _ in self._info.variables.items():
+                try:
+                    self._attributes[name] = self._info.__getattr__(name)
+                except Exception as ex:  # pylint: disable=broad-except
```
Same as above.

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    def update(self):
+        """Get the latest data from the Particle Cloud for this device."""
+        if self._info.variables is not None:
+            for name, _ in self._info.variables.items():
```
You do not need `.items()` if you only use the key:
```suggestion
            for name in self._info.variables:
```

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
+
+    devices = {}
```
For what is `devices` used?

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    device_list = []
+
+    for name, info in particle.devices.items():
+        entity_id = f"{DOMAIN}.{name.lower()}"
```
For what is `entity_id` used? What kind of [entity](https://developers.home-assistant.io/docs/en/entity_index.html) should this integration create?

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+
+DOMAIN = "particle"
+
+_LOGGER = logging.getLogger(DOMAIN)
```
Please use __name__ and move it above the constant `DOMAIN`:
```suggestion
_LOGGER = logging.getLogger(__name__)
```

springstan commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+        return False
+
+    component.async_register_entity_service(
+        "call", FUNCTION_CALL_SCHEMA, ParticleDevice.call
```
Please create a constant for `"call"`.

Schoonology commented:
Commented above on a specific line of code, but it's also applicable to the entire PR.

MartinHjelmare commented:
@Schoonology will you address the review comments and work to get the PR merged? Otherwise I suggest we close here now.

Schoonology commented:
@MartinHjelmare Thanks for checking. I still intend to address the review comments, if there's still an interest in this feature existing.

MartinHjelmare commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+        _LOGGER.error("Unable to connect to Particle Cloud: %s", str(ex))
+        return False
+
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
```
Let's don't do this and instead create separate sensor entities in a sensor platform as needed for each attribute. We can load other platforms via discovery, discovery.async_load_platform. See other integrations for examples.

We can register a service here in the base integration module that calls the cloud function and takes the name of the device as service data.

Schoonology replied:
How's the documentation for doing this refactor? This has been working for quite some time, and it seems like there's been some internal, architectural changes that are affecting the way integrations are meant to be written.

It took me a long time to get this version working and approved, as:

- Updates take a long time to test, particularly for a DIY device like Particles.
- The desired output is a moving target.
- Documentation disagreed on what should be written where, and what documentation was considered "blessed" disagreed with the code already running on the contemporary versions of Home Assistant.

Your question earlier today was whether I intended to work through the review comments. I really do, but it's a lot to work through for a small, free contribution I wrote 6 months ago.

What I want has remained the same: I want some idea that when I'm done working through the current comments, I'll be done with this PR and it can be merged.

This started out simple enough: I had a custom component for a popular DIY platform, and I thought I'd contribute it to other Home Assistant users. If you have a moment to consider what this feels like for a newcomer, consider that every message I've received has included desired changes that conflict with what official documentation I could find, and some of which conflict with other reviews I've received on the same code.

So, once more from the top: where can I go to figure out what I _actually_ need to do here?

MartinHjelmare replied:
Read our developer documentation and look at existing examples that are recently added. Older integrations are often not up to standard.

I can't give any guarantees on how much feedback there will be. We have standards that we want to maintain and we won't merge a PR until we think that it's ok according to those standards.

From yesterday we require all new integrations for devices and services to use a config flow and config entries. For the PRs where review has started we are making exceptions on this rule. That's why I haven't said anything about that until now.

I'd recommend reading the whole section about creating integrations that start with the file structure:
https://developers.home-assistant.io/docs/creating_integration_file_structure

Here's the docs on how to load a platform via discovery:
https://developers.home-assistant.io/docs/creating_component_generic_discovery#loading-platforms-when-configured-via-configurationyaml

MartinHjelmare commented on homeassistant/components/particle/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    @property
+    def state(self):
+        """Return a high-level state for the device."""
+        if self._info.connected:
```
The connected info feature should be a binary sensor entity in a binary sensor platform with device class `"connection"`.
